lib/ProductOpener/Ingredients.pm

```diff
@@ -6790,33 +6790,55 @@ The match function takes 2 arguments:
 Used to compute % of fruits and vegetables, % of milk etc. which is needed by some algorithm
 like the Nutri-Score.
 
+=head3 Return values
+
+=head4 $percent
+
+Sum of matching ingredients percent.
+
+=head4 $water_percent
+
+Percent of water (used to recompute the percentage for categories of products that are consumed after removing water)
+
 =cut
 
 sub add_ingredients_matching_function ($ingredients_ref, $match_function_ref) {
 
-	my $count = 0;
+	my $percent = 0;
+	my $water_percent = 0;
 
 	foreach my $ingredient_ref (@{$ingredients_ref}) {
+
+		my $ingredient_percent;
+		if (defined $ingredient_ref->{percent}) {
+			$ingredient_percent = $ingredient_ref->{percent};
+		}
+		elsif (defined $ingredient_ref->{percent_estimate}) {
+			$ingredient_percent = $ingredient_ref->{percent_estimate};
+		}
+
 		my $match = $match_function_ref->($ingredient_ref->{id}, $ingredient_ref->{processing});
-		if ($match) {
-			if (defined $ingredient_ref->{percent}) {
-				$count += $ingredient_ref->{percent};
-			}
-			elsif (defined $ingredient_ref->{percent_estimate}) {
-				$count += $ingredient_ref->{percent_estimate};
-			}
-			# We may not have percent_min if the ingredient analysis failed because of seemingly impossible values
-			# in that case, try to get the possible percent values in nested sub ingredients
-			elsif (defined $ingredient_ref->{ingredients}) {
-				$count += add_ingredients_matching_function($ingredient_ref->{ingredients}, $match_function_ref);
-			}
+
+		# We have a match and a percent
+		if (($match) and (defined $ingredient_percent)) {
+			$percent += $ingredient_percent;
 		}
+		# The ingredient does not match,
+		# or we don't have percent_estimate if the ingredient analysis failed because of seemingly impossible values
+		# in that case, try to get the possible percent values in nested sub ingredients
 		elsif (defined $ingredient_ref->{ingredients}) {
-			$count += add_ingredients_matching_function($ingredient_ref->{ingredients}, $match_function_ref);
+			my ($sub_ingredients_percent, $sub_ingredients_water_percent)
+				= add_ingredients_matching_function($ingredient_ref->{ingredients}, $match_function_ref);
+			$percent += $sub_ingredients_percent;
+			$water_percent += $sub_ingredients_water_percent;
+		}
+		# Keep track of water
+		elsif (is_a("ingredients", $ingredient_ref->{id}, 'en:water')) {
```

Comment:
I'm not sure why it's an elsif here ?

Are we sure that water never match ? Why not a simple if ? Or, maybe, put it as first condition ?

Reply:
It was avoid double counting water if we have something like "flavoured water (water, flavour)", but I'll do it differently.

Reply:
A comment maybe enough then